File: python/src/ttio/raman_image.py
"""``RamanImage`` — Raman hyperspectral imaging cube."""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import ClassVar

import numpy as np

from .enums import ImageKind, SpectralAxisKind
from .image import Image


def _parse_double_attr(value) -> float:
    """Parse a double attribute that may be a float/int (native) or str (legacy)."""
    if isinstance(value, (int, float)):
        return float(value)
    return float(str(value))



@dataclass(slots=True)
class RamanImage(Image):
    """Raman hyperspectral imaging dataset: a ``width × height`` grid of
    pixels, each pixel a spectrum of ``spectral_points`` intensity values
    sampled at a shared rank-1 ``wavenumbers`` axis.

    Parameters
    ----------
    width, height : int, default 0
        Image grid dimensions in pixels.
    spectral_points : int, default 0
        Number of float64 intensity values per pixel.
    intensity : numpy.ndarray, default empty rank-3 array
        Rank-3 intensity cube of shape ``(height, width, spectral_points)``.
    wavenumbers : numpy.ndarray, default empty rank-1 array
        Rank-1 wavenumber axis of length ``spectral_points`` (1/cm).
    pixel_size_x, pixel_size_y : float, default 0.0
        Spatial pixel size (implementation-defined units).
    scan_pattern : str, default ""
        Scan pattern label (``"raster"``, ``"random_access"``, ...).
    tile_size : int, default 0
        HDF5 chunk tile size for reads; 0 means native cube chunks.
    excitation_wavelength_nm : float, default 0.0
        Laser excitation wavelength in nanometres.
    laser_power_mw : float, default 0.0
        Incident laser power in milliwatts.
    title, isa_investigation_id : str, default ""
        Dataset-level metadata.
    identifications, quantifications, provenance_records : list, default []
        Dataset-level composition fields (ObjC inherits these from
        ``TTIOSpectralDataset``).

    Notes
    -----
    API status: Stable (v0.11, M73).

    Cross-language equivalents
    --------------------------
    Objective-C: ``TTIORamanImage`` · Java:
    ``global.thalion.ttio.RamanImage``.
    """

    # Common fields are inherited from Image; only Raman-specific fields here.
    wavenumbers: np.ndarray = field(default_factory=lambda: np.zeros((0,)))
    excitation_wavelength_nm: float = 0.0
    laser_power_mw: float = 0.0

    kind: ClassVar[ImageKind] = ImageKind.RAMAN

    @property
    def spectral_axis(self) -> np.ndarray:
        """The wavenumber axis (alias of :attr:`wavenumbers`)."""
        return self.wavenumbers

    @property
    def spectral_axis_kind(self) -> SpectralAxisKind:
        return SpectralAxisKind.WAVENUMBER

    def __post_init__(self) -> None:
        if self.width == 0 and self.height == 0 and self.spectral_points == 0:
            return
        if self.intensity.ndim != 3:
            raise ValueError(
                f"intensity must be rank-3, got shape={self.intensity.shape}"
            )
        h, w, sp = self.intensity.shape
        if (h, w, sp) != (self.height, self.width, self.spectral_points):
            raise ValueError(
                f"intensity shape {(h, w, sp)} does not match "
                f"(height, width, spectral_points)="
                f"{(self.height, self.width, self.spectral_points)}"
            )
        if self.wavenumbers.ndim != 1 or self.wavenumbers.shape[0] != self.spectral_points:
            raise ValueError(
                f"wavenumbers shape {self.wavenumbers.shape} does not match "
                f"spectral_points={self.spectral_points}"
            )


    def write_to(self, study_group) -> None:
        """Write this Raman image cube under ``<study_group>/raman_image_cube/``.

        Mirrors :meth:`global.thalion.ttio.RamanImage.writeTo` — intensity
        as a 3-D ``[h, w, sp]`` dataset, ``wavenumbers`` 1-D dataset,
        and Raman-specific scalar attributes.

        Pixel size attributes are written as **native double** (not strings),
        applying the PR #31 lesson immediately so legacy-string reads remain
        backwards-compatible via :meth:`read_from`.
        """
        from ttio.enums import Compression, Precision
        ic = study_group.create_group("raman_image_cube")
        ic.set_attribute("width", int(self.width))
        ic.set_attribute("height", int(self.height))
        ic.set_attribute("spectral_points", int(self.spectral_points))
        ic.set_attribute("pixel_size_x", float(self.pixel_size_x))
        ic.set_attribute("pixel_size_y", float(self.pixel_size_y))
        ic.set_attribute("scan_pattern", self.scan_pattern)
        ic.set_attribute("excitation_wavelength_nm", float(self.excitation_wavelength_nm))
        ic.set_attribute("laser_power_mw", float(self.laser_power_mw))

        intensity_ds = ic.create_dataset_nd(
            "intensity", Precision.FLOAT64,
            shape=(self.height, self.width, self.spectral_points),
            chunks=(1, 1, self.spectral_points),
            compression=Compression.ZLIB, compression_level=6,
        )
        intensity_ds.write(np.ascontiguousarray(self.intensity, dtype=np.float64))

        wn_ds = ic.create_dataset_nd(
            "wavenumbers", Precision.FLOAT64,
            shape=(self.spectral_points,),
            chunks=(self.spectral_points,),
            compression=Compression.ZLIB, compression_level=6,
        )
        wn_ds.write(np.ascontiguousarray(self.wavenumbers, dtype=np.float64))

    @classmethod
    def read_from(cls, study_group) -> "RamanImage | None":
        """Read a Raman image cube from a study group; return None if absent.

        Accepts both the native-double form (written by this class) and
        the legacy string form for pixel_size_x/y,
        excitation_wavelength_nm, and laser_power_mw.
        """
        if not study_group.has_child("raman_image_cube"):
            return None
        ic = study_group.open_group("raman_image_cube")
        width = int(ic.get_attribute("width"))
        height = int(ic.get_attribute("height"))
        spectral_points = int(ic.get_attribute("spectral_points"))
        pixel_size_x = (_parse_double_attr(ic.get_attribute("pixel_size_x"))
                        if ic.has_attribute("pixel_size_x") else 0.0)
        pixel_size_y = (_parse_double_attr(ic.get_attribute("pixel_size_y"))
                        if ic.has_attribute("pixel_size_y") else 0.0)
        scan_pattern = (ic.get_attribute("scan_pattern")
                        if ic.has_attribute("scan_pattern") else "")
        excitation_wavelength_nm = (
            _parse_double_attr(ic.get_attribute("excitation_wavelength_nm"))
            if ic.has_attribute("excitation_wavelength_nm") else 0.0
        )
        laser_power_mw = (
            _parse_double_attr(ic.get_attribute("laser_power_mw"))
            if ic.has_attribute("laser_power_mw") else 0.0
        )

        intensity_raw = np.asarray(ic.open_dataset("intensity").read())
        intensity = intensity_raw.reshape(height, width, spectral_points)
        wavenumbers = np.asarray(ic.open_dataset("wavenumbers").read(), dtype=np.float64)

        return cls(
            width=width, height=height, spectral_points=spectral_points,
            pixel_size_x=pixel_size_x, pixel_size_y=pixel_size_y,
            intensity=intensity, wavenumbers=wavenumbers,
            scan_pattern=scan_pattern,
            excitation_wavelength_nm=excitation_wavelength_nm,
            laser_power_mw=laser_power_mw,
        )


__all__ = ["RamanImage"]

File: java/src/main/java/global/thalion/ttio/RamanImage.java
/*
 * TTI-O Java Implementation
 * Copyright (c) 2026 The Thalion Initiative
 * SPDX-License-Identifier: LGPL-3.0-or-later
 */
package global.thalion.ttio;

import global.thalion.ttio.Enums.Compression;
import global.thalion.ttio.Enums.ImageKind;
import global.thalion.ttio.Enums.Precision;
import global.thalion.ttio.Enums.SpectralAxisKind;
import global.thalion.ttio.providers.StorageDataset;
import global.thalion.ttio.providers.StorageGroup;

import java.util.List;

/**
 * Raman hyperspectral imaging dataset: {@code width × height} pixel
 * grid, each pixel a Raman spectrum of {@code spectralPoints}
 * intensity values sampled at a shared wavenumber axis.
 *
 * <p>Stored under {@code /study/raman_image_cube/} as a 3-D
 * float64 intensity cube with a 1-D {@code wavenumbers} axis.</p>
 *
 * <p>Composition vs inheritance notes match {@link MSImage}.</p>
 *
 * <p><b>API status:</b> Stable (v0.11, M73).</p>
 *
 * <p><b>Cross-language equivalents:</b> Objective-C
 * {@code TTIORamanImage}, Python
 * {@code ttio.raman_image.RamanImage}.</p>
 *
 *
 */
public class RamanImage extends Image {

    private static final String GROUP_NAME = "raman_image_cube";

    private final double excitationWavelengthNm;
    private final double laserPowerMw;
    private final double[] wavenumbers;

    /**
     * Full constructor capturing every cube dimension, laser metadata,
     * and study-level annotation.
     *
     * @param width                   image width in pixels
     * @param height                  image height in pixels
     * @param spectralPoints          points along the Raman shift axis
     * @param tileSize                HDF5 chunk side; 0 means full-image
     *                                chunking
     * @param pixelSizeX              pixel pitch in micrometres along X
     * @param pixelSizeY              pixel pitch in micrometres along Y
     * @param scanPattern             free-form scan-pattern label
     * @param excitationWavelengthNm  laser excitation wavelength in nm
     * @param laserPowerMw            laser power at the sample in mW
     * @param intensityCube           row-major flat cube of shape
     *                                {@code [height, width, spectralPoints]}
     * @param wavenumbers             per-band Raman-shift values
     * @param title                   study title; null becomes empty
     * @param isaInvestigationId      ISA investigation identifier; null
     *                                becomes empty
     * @param identifications         dataset-level identifications;
     *                                null becomes empty
     * @param quantifications         dataset-level quantifications;
     *                                null becomes empty
     * @param provenanceRecords       dataset-level provenance chain;
     *                                null becomes empty
     */
    public RamanImage(int width, int height, int spectralPoints, int tileSize,
                      double pixelSizeX, double pixelSizeY, String scanPattern,
                      double excitationWavelengthNm, double laserPowerMw,
                      double[] intensityCube, double[] wavenumbers,
                      String title, String isaInvestigationId,
                      List<Identification> identifications,
                      List<Quantification> quantifications,
                      List<ProvenanceRecord> provenanceRecords) {
        super(width, height, spectralPoints, tileSize,
              pixelSizeX, pixelSizeY, scanPattern, intensityCube,
              title, isaInvestigationId,
              identifications, quantifications, provenanceRecords);
        this.excitationWavelengthNm = excitationWavelengthNm;
        this.laserPowerMw = laserPowerMw;
        this.wavenumbers = wavenumbers;
    }

    /** Convenience — image-only construction (empty dataset-level metadata). */
    public RamanImage(int width, int height, int spectralPoints,
                      double pixelSizeX, double pixelSizeY, String scanPattern,
                      double excitationWavelengthNm, double laserPowerMw,
                      double[] intensityCube, double[] wavenumbers) {
        this(width, height, spectralPoints, 0,
             pixelSizeX, pixelSizeY, scanPattern,
             excitationWavelengthNm, laserPowerMw,
             intensityCube, wavenumbers,
             "", "", List.of(), List.of(), List.of());
    }

    /** @return Laser excitation wavelength in nm. */
    public double excitationWavelengthNm() { return excitationWavelengthNm; }

    /** @return Laser power at the sample in mW. */
    public double laserPowerMw() { return laserPowerMw; }

    /** @return Per-band Raman-shift values in {@code cm^-1}. */
    public double[] wavenumbers() { return wavenumbers; }

    @Override
    public ImageKind kind() { return ImageKind.RAMAN; }

    @Override
    public double[] spectralAxis() { return wavenumbers; }

    @Override
    public SpectralAxisKind spectralAxisKind() { return SpectralAxisKind.WAVENUMBER; }

    /** Write this image cube as an HDF5 sub-group of {@code studyGroup}. */
    public void writeTo(StorageGroup studyGroup) {
        try (StorageGroup ic = studyGroup.createGroup(GROUP_NAME)) {
            ic.setAttribute("width", (long) width);
            ic.setAttribute("height", (long) height);
            ic.setAttribute("spectral_points", (long) spectralPoints);
            ic.setAttribute("pixel_size_x", Double.valueOf(pixelSizeX));
            ic.setAttribute("pixel_size_y", Double.valueOf(pixelSizeY));
            ic.setAttribute("excitation_wavelength_nm", Double.valueOf(excitationWavelengthNm));
            ic.setAttribute("laser_power_mw", Double.valueOf(laserPowerMw));
            if (scanPattern != null)
                ic.setAttribute("scan_pattern", scanPattern);
            if (tileSize > 0)
                ic.setAttribute("tile_size", (long) tileSize);

            long chunkSize = tileSize > 0 ? tileSize : 1;
            long[] shape = { height, width, spectralPoints };
            long[] chunks = { chunkSize, chunkSize, spectralPoints };
            try (StorageDataset ds = ic.createDatasetND("intensity",
                    Precision.FLOAT64, shape, chunks,
                    Compression.ZLIB, 6)) {
                ds.writeAll(intensityCube);
            }

            long[] axisShape = { spectralPoints };
            long[] axisChunks = { spectralPoints };
            try (StorageDataset wn = ic.createDatasetND("wavenumbers",
                    Precision.FLOAT64, axisShape, axisChunks,
                    Compression.ZLIB, 6)) {
                wn.writeAll(wavenumbers);
            }
        }
    }

    /** Read a Raman image cube from a study group, or {@code null} if absent. */
    public static RamanImage readFrom(StorageGroup studyGroup) {
        if (!studyGroup.hasChild(GROUP_NAME)) return null;
        try (StorageGroup ic = studyGroup.openGroup(GROUP_NAME)) {
            int width = ((Number) ic.getAttribute("width")).intValue();
            int height = ((Number) ic.getAttribute("height")).intValue();
            int spectralPoints = ((Number) ic.getAttribute("spectral_points")).intValue();
            double pixelSizeX = ic.hasAttribute("pixel_size_x")
                    ? parseDoubleAttr(ic.getAttribute("pixel_size_x")) : 0.0;
            double pixelSizeY = ic.hasAttribute("pixel_size_y")
                    ? parseDoubleAttr(ic.getAttribute("pixel_size_y")) : 0.0;
            double excitationWavelengthNm = ic.hasAttribute("excitation_wavelength_nm")
                    ? parseDoubleAttr(ic.getAttribute("excitation_wavelength_nm")) : 0.0;
            double laserPowerMw = ic.hasAttribute("laser_power_mw")
                    ? parseDoubleAttr(ic.getAttribute("laser_power_mw")) : 0.0;
            String scanPattern = ic.hasAttribute("scan_pattern")
                    ? (String) ic.getAttribute("scan_pattern") : null;
            int tileSize = ic.hasAttribute("tile_size")
                    ? ((Number) ic.getAttribute("tile_size")).intValue() : 0;

            double[] cube;
            try (StorageDataset ds = ic.openDataset("intensity")) {
                cube = (double[]) ds.readAll();
            }
            double[] wn;
            try (StorageDataset ds = ic.openDataset("wavenumbers")) {
                wn = (double[]) ds.readAll();
            }
            return new RamanImage(width, height, spectralPoints, tileSize,
                    pixelSizeX, pixelSizeY, scanPattern,
                    excitationWavelengthNm, laserPowerMw,
                    cube, wn,
                    "", "", List.of(), List.of(), List.of());
        }
    }

    /**
     * Parse a double attribute that may be stored as either a
     * {@link Number} (new native-double form) or a {@link String}
     * (legacy form written before the PR #31 fix).
     *
     * <p>Copy-pasted from {@code MSImage} deliberately: the helper is
     * small, and cross-class coupling via a shared utility would
     * require widening the package-private API surface without
     * meaningful benefit.</p>
     */
    private static double parseDoubleAttr(Object value) {
        if (value instanceof Number n) return n.doubleValue();
        return Double.parseDouble((String) value);
    }
}